src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -799,29 +799,26 @@ void ScriptEngine::DoAutoReloadPluginCheck()
 
 void ScriptEngine::AutoReloadPlugins()
 {
-    if (_changedPluginFiles.size() > 0)
+    std::lock_guard guard(_changedPluginFilesMutex);
+    for (const auto& path : _changedPluginFiles)
     {
-        std::lock_guard<std::mutex> guard(_changedPluginFilesMutex);
-        for (auto& path : _changedPluginFiles)
+        auto findResult = std::find_if(_plugins.begin(), _plugins.end(), [&path](const std::shared_ptr<Plugin>& plugin) {
+            return Path::Equals(path, plugin->GetPath());
+        });
+        if (findResult != _plugins.end())
         {
-            auto findResult = std::find_if(_plugins.begin(), _plugins.end(), [&path](const std::shared_ptr<Plugin>& plugin) {
-                return Path::Equals(path, plugin->GetPath());
-            });
-            if (findResult != _plugins.end())
+            auto& plugin = *findResult;
+            try
             {
-                auto& plugin = *findResult;
-                try
-                {
-                    ReloadPlugin(plugin);
-                }
-                catch (const std::exception& e)
-                {
-                    _console.WriteLineError(e.what());
-                }
+                ReloadPlugin(plugin);
+            }
+            catch (const std::exception& e)
+            {
+                _console.WriteLineError(e.what());
             }
         }
-        _changedPluginFiles.clear();
     }
+    _changedPluginFiles.clear();
```

Comment:
What is the reason behind this change?

Reply:
> Made ScriptEngine::AutoReloadPlugins() take a mutex earlier as checking if the container is not empty while not holding a mutex is probably a bad idea too.

This is a diff from moving everything a single indent to the left by removing `if (_changedPluginFiles.size() > 0)`, as a result all operations on the vector are under a mutex.

Reply:
The reason it checked if the vector was not empty first was to improve performance. It would only lock if needed to enumerate the vector. Reading the size of the vector should be a thread safe op.

Reply:
It may change between reading the length and accessing it, so even if reading the length of the vector is atomic, it's not thread safe.

Reply:
I don't think it should be a concern here - this check is done only once per frame, and mutexes are very fast to lock when not under contention. Even if it's a thread safe operation, you risk unneccessary race conditions, albeit harmless ones.

Reply:
> It may change between reading the length and accessing it, so even if reading the length of the vector is atomic, it's not thread safe.

That is what I meant, if it is atomic that is fine. It just means that it won't process the loop until next time round. This is to reduce frequeny of a lock.

Reply:
It's still negligible, one uncontented lock per frame is nothing.

Reply:
That was my reason for putting it there in the first place, is there a good reason for removing other than "it is negligible"? It is +1 lock per frame.

Reply:
> is there a good reason for removing other than "it is negligible"?

Correctness

Reply:
I agree with CookiePLMonster that it's better to avoid any data racing than to assume that the next frame the size will match.

Reply:
Okay, I personally disagree, I have always implemented this concept for both personal projects, OpenRCT2 and work.

Reply:
I can skip this entire function, and thus the mutex, if `_plugins.empty()` since if that is the case then the for loop will never do any action anyway - is that a reasonable compromise?

Reply:
I would have no problem approving it if it were just the changes for the fix. The string_view is fine as well, although @Broxzier did point out the overlap with another PR.

Reply:
Don't mind the other PR, I think that one should be made smaller anyway.

Reply:
OK, I reintroduced the emptiness check before taking the mutex so this PR doesn't stay unmerged.